# Python言語によるビジネスアナリティクス

## 実務家のための最適化，統計分析，機械学習（近代科学社）


## scikit-learn を用いた 機械学習の例題と練習問題 


## 教科書 付録 B参照

### B.2 線形回帰

広告のデータ http://logopt.com/data/Advertising.csv を用いる．

テレビ(TV)，ラジオ(Radio)，新聞(Newspaper)への広告から売り上げ(Sales)を予測する．


In [2]:
import pandas as pd #まずはpandasモジュールを準備する．

In [3]:
#csvファイルからデータ読み込み
data = pd.read_csv('http://logopt.com/data/Advertising.csv', index_col =0) #0行目をインデックスにする．
data.head()

TV  Radio  Newspaper  Sales
1  230.1   37.8       69.2   22.1
2   44.5   39.3       45.1   10.4
3   17.2   45.9       69.3    9.3
4  151.5   41.3       58.5   18.5
5  180.8   10.8       58.4   12.9

独立変数（特徴ベクトル）$X$ は TV, Ratio, Newspaperの列，従属変数 $y$ は Salesの列


In [3]:
X = data[['TV', 'Radio', 'Newspaper']]
y = data['Sales']
X.head()

TV  Radio  Newspaper
1  230.1   37.8       69.2
2   44.5   39.3       45.1
3   17.2   45.9       69.3
4  151.5   41.3       58.5
5  180.8   10.8       58.4

## scikit-learnの基本手順

### 手順１：クラスをインポートして，インスタンスを生成する．

### 手順２：fitメソッドを用いて，データから学習する．

### 手順３：predictメソッドを用いて予測を行う．


In [1]:
from sklearn.linear_model import LinearRegression #線形回帰クラス LinearRegression のインポート
reg = LinearRegression() # 線形回帰クラスのインスタンス reg 生成
reg.fit(X, y)           # fitによる学習
yhat = reg.predict(X)   # predictによる予測

NameError: name 'X' is not defined

In [5]:
print("y-切片= ",reg.intercept_)
print("係数 = ",reg.coef_)

y-切片=  2.9388893694594067
係数 =  [ 0.04576465  0.18853002 -0.00103749]


In [6]:
SSE = ((yhat-y)**2).sum()      # Sum of Square Error
SST = ((y.mean()-y)**2).sum()  # Total Sum of Square
print("R2 =", 1-SSE/SST)       # 決定係数 R^2

R2 = 0.8972106381789522


In [7]:
print(reg.score(X,y)) #決定係数の別計算

0.8972106381789521


## 問題

http://logopt.com/data/SATGPA.csv データを用いて，2種類のSATの成績からGPAを予測せよ．

In [8]:
#csvファイルからデータ読み込み
data = pd.read_csv('http://logopt.com/data/SATGPA.csv', index_col =0) #0行目をインデックスにする．
data.head()

MathSAT  VerbalSAT   GPA
1      580        420  2.90
2      670        530  2.83
3      680        540  2.90
4      630        640  3.30
5      620        630  3.61

In [9]:
X = data[['MathSAT', 'VerbalSAT']]
y = data['GPA']
X.head()

MathSAT  VerbalSAT
1      580        420
2      670        530
3      680        540
4      630        640
5      620        630

In [10]:
from sklearn.linear_model import LinearRegression #線形回帰クラス LinearRegression のインポート
reg = LinearRegression() # 線形回帰クラスのインスタンス reg 生成
reg.fit(X, y)           # fitによる学習
yhat = reg.predict(X)   # predictによる予測

In [11]:
print("y-切片= ",reg.intercept_)
print("係数 = ",reg.coef_)

y-切片=  2.7374715623718373
係数 =  [-0.00037958  0.00108557]


In [12]:
SSE = ((yhat-y)**2).sum()      # Sum of Square Error
SST = ((y.mean()-y)**2).sum()  # Total Sum of Square
print("R2 =", 1-SSE/SST)       # 決定係数 R^2

R2 = 0.06645943942486443


In [13]:
print(reg.score(X,y)) #決定係数の別計算

0.06645943942486465


### 問題

http://logopt.com/data/Boston.csv のBostonの住宅データを用いて回帰分析を行え．

medvが住宅の価格で，他のデータ（犯罪率や人口など）から予測する．


In [14]:
#csvファイルからデータ読み込み
data = pd.read_csv('http://logopt.com/data/Boston.csv', index_col =0) #0行目をインデックスにする．
data.head()

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
1  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
2  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
3  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
4  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
5  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

    black  lstat  medv  
1  396.90   4.98  24.0  
2  396.90   9.14  21.6  
3  392.83   4.03  34.7  
4  394.63   2.94  33.4  
5  396.90   5.33  36.2

In [16]:
X = data[['crim', 'zn','indus','tax']]
y = data['medv']
X.head()

crim    zn  indus  tax
1  0.00632  18.0   2.31  296
2  0.02731   0.0   7.07  242
3  0.02729   0.0   7.07  242
4  0.03237   0.0   2.18  222
5  0.06905   0.0   2.18  222

In [18]:
from sklearn.linear_model import LinearRegression #線形回帰クラス LinearRegression のインポート
reg = LinearRegression() # 線形回帰クラスのインスタンス reg 生成
reg.fit(X, y)           # fitによる学習
yhat = reg.predict(X)   # predictによる予測
SE = ((yhat-y)**2).sum()      # Sum of Square Error
SST = ((y.mean()-y)**2).sum()  # Total Sum of Square
print("R2 =", 1-SSE/SST)       # 決定係数 R^2

R2 = 0.9999491324334212


In [19]:
print(reg.score(X,y)) #決定係数の別計算

0.30429948592131517


### 問題

http://logopt.com/data/Auto.csv の車の燃費のデータを用いて回帰分析を行え．

データの詳細については，

https://vincentarelbundock.github.io/Rdatasets/doc/ISLR/Auto.html

を参照せよ．

最初の列が燃費（mpg: Mile Per Gallon)であり，これを他の列の情報を用いて予測する．最後の列は車名なので無視して良い．

In [22]:
#csvファイルからデータ読み込み
data = pd.read_csv('http://logopt.com/data/Auto.csv', index_col =0) #0行目をインデックスにする．
data.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0  18.0          8         307.0         130    3504          12.0    70   
1  15.0          8         350.0         165    3693          11.5    70   
2  18.0          8         318.0         150    3436          11.0    70   
3  16.0          8         304.0         150    3433          12.0    70   
4  17.0          8         302.0         140    3449          10.5    70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [23]:
X = data[['cylinders','displacement','horsepower','weight','acceleration']]
y = data['mpg']
X.head()

cylinders  displacement  horsepower  weight  acceleration
0          8         307.0         130    3504          12.0
1          8         350.0         165    3693          11.5
2          8         318.0         150    3436          11.0
3          8         304.0         150    3433          12.0
4          8         302.0         140    3449          10.5

In [24]:
from sklearn.linear_model import LinearRegression #線形回帰クラス LinearRegression のインポート
reg = LinearRegression() # 線形回帰クラスのインスタンス reg 生成
reg.fit(X, y)           # fitによる学習
yhat = reg.predict(X)   # predictによる予測
SE = ((yhat-y)**2).sum()      # Sum of Square Error
SST = ((y.mean()-y)**2).sum()  # Total Sum of Square
print("R2 =", 1-SSE/SST)       # 決定係数 R^2

R2 = 0.999908775574257


In [25]:
print(reg.score(X,y)) #決定係数の別計算

0.7076926326770515


### 例題２

http://logopt.com/data/Diamond.csv からダイアモンドの価格データを読み込み，線形回帰による予測を行う．

列は ["carat","colour","clarity","certification","price"] であり，他の情報から価格(price)の予測を行え．

カラット(carat)以外の列は情報が文字列として保管されている．

これはカテゴリー変数とよばれ，sciki-learnで扱うには，数値に変換してあげる必要がある．

pandasの**get_dummies関数**で数値情報（ダミー変数）に変換してから，線形回帰を行う．

たとえば，色を表すcolour列はD,E,F,G,H,Iの文字列が入っている．これを各値が入っているとき1，それ以外のとき0の数値に変換したものが**ダミー変数**になる．

色はいずれかの値をとるので，ダミー変数は独立でない（1つが1になると，他のすべては0になる）．

最初のダミー変数を除くには，get_dummies関数の引数の**drop_first**をTrueに設定すれば良い．

In [30]:
diamond = pd.read_csv('http://logopt.com/data/Diamond.csv', index_col=0)
diamond.head()

carat colour clarity certification  price
1   0.30      D     VS2           GIA   1302
2   0.30      E     VS1           GIA   1510
3   0.30      G    VVS1           GIA   1510
4   0.30      G     VS1           GIA   1260
5   0.31      D     VS1           GIA   1641

In [31]:
diamond = pd.get_dummies(diamond, drop_first=True) #ダミー変数の最初のものを除く
#diamond = pd.get_dummies(diamond) # 除かなくても結果は同じ
diamond.head()

carat  price  colour_E  colour_F  colour_G  colour_H  colour_I  \
1   0.30   1302         0         0         0         0         0   
2   0.30   1510         1         0         0         0         0   
3   0.30   1510         0         0         1         0         0   
4   0.30   1260         0         0         1         0         0   
5   0.31   1641         0         0         0         0         0   

   clarity_VS1  clarity_VS2  clarity_VVS1  clarity_VVS2  certification_HRD  \
1            0            1             0             0                  0   
2            1            0             0             0                  0   
3            0            0             1             0                  0   
4            1            0             0             0                  0   
5            1            0             0             0                  0   

   certification_IGI  
1                  0  
2                  0  
3                  0  
4                  0  
5                  0

In [32]:
y = diamond.price                # 従属変数(price)の抽出
X = diamond.drop('price',axis=1) # 独立変数（特徴ベクトル）をpriceの列を除くことによって生成
X.head()

carat  colour_E  colour_F  colour_G  colour_H  colour_I  clarity_VS1  \
1   0.30         0         0         0         0         0            0   
2   0.30         1         0         0         0         0            1   
3   0.30         0         0         1         0         0            0   
4   0.30         0         0         1         0         0            1   
5   0.31         0         0         0         0         0            1   

   clarity_VS2  clarity_VVS1  clarity_VVS2  certification_HRD  \
1            1             0             0                  0   
2            0             0             0                  0   
3            0             1             0                  0   
4            0             0             0                  0   
5            0             0             0                  0   

   certification_IGI  
1                  0  
2                  0  
3                  0  
4                  0  
5                  0

In [33]:
from sklearn.linear_model import LinearRegression #線形回帰クラスのインポート
reg = LinearRegression() # 線形回帰クラスのインスタンス生成
reg.fit(X, y)           # 学習
yhat = reg.predict(X)   # 予測

In [34]:
print("y-切片= ",reg.intercept_)
print("係数 = ",reg.coef_)
print("決定変数= ",reg.score(X,y)) #決定係数の別計算

y-切片=  169.17604383491016
係数 =  [12766.39597047 -1439.0853427  -1841.69054716 -2176.67218633
 -2747.14998002 -3313.1023993  -1474.56614749 -1792.01092358
  -689.29043537 -1191.16426364    15.22672874   141.2624469 ]
決定変数=  0.9581280577870392


### 問題

http://logopt.com/data/carprice.csv から車の価格データを読み込み，線形回帰による予測を行え．

データの詳細は
https://vincentarelbundock.github.io/Rdatasets/doc/DAAG/carprice.html
にある．

車種(Type)，100マイル走る際のガロン数（gpm100），都市部での1ガロンあたりの走行距離（MPG.city），高速道路での１ガロン当たりの走行距離（MPG.highway）から，価格(Price)を予測せよ．

In [35]:
cardate = pd.read_csv('http://logopt.com/data/carprice.csv', index_col=0)
cardate.head()

Type  MinPrice  Price  MaxPrice  RangePrice  RoughRange  gpm100  \
6   Midsize      14.2   15.7      17.3         3.1        3.09     3.8   
7     Large      19.9   20.8      21.7         1.8        1.79     4.2   
8     Large      22.6   23.7      24.9         2.3        2.31     4.9   
9   Midsize      26.3   26.3      26.3         0.0       -0.01     4.3   
10    Large      33.0   34.7      36.3         3.3        3.30     4.9   

    MPGcity  MPGhighway  
6        22          31  
7        19          28  
8        16          25  
9        19          27  
10       16          25

In [36]:
cardate = pd.get_dummies(cardate, drop_first=True) #ダミー変数の最初のものを除く
cardate.head()

MinPrice  Price  MaxPrice  RangePrice  RoughRange  gpm100  MPGcity  \
6       14.2   15.7      17.3         3.1        3.09     3.8       22   
7       19.9   20.8      21.7         1.8        1.79     4.2       19   
8       22.6   23.7      24.9         2.3        2.31     4.9       16   
9       26.3   26.3      26.3         0.0       -0.01     4.3       19   
10      33.0   34.7      36.3         3.3        3.30     4.9       16   

    MPGhighway  Type_Large  Type_Midsize  Type_Small  Type_Sporty  Type_Van  
6           31           0             1           0            0         0  
7           28           1             0           0            0         0  
8           25           1             0           0            0         0  
9           27           0             1           0            0         0  
10          25           1             0           0            0         0

In [41]:
y = cardate.Price                # 従属変数(price)の抽出
X = cardate.drop('Price',axis=1) # 独立変数（特徴ベクトル）をpriceの列を除くことによって生成
X.head()

MinPrice  MaxPrice  RangePrice  RoughRange  gpm100  MPGcity  MPGhighway  \
6       14.2      17.3         3.1        3.09     3.8       22          31   
7       19.9      21.7         1.8        1.79     4.2       19          28   
8       22.6      24.9         2.3        2.31     4.9       16          25   
9       26.3      26.3         0.0       -0.01     4.3       19          27   
10      33.0      36.3         3.3        3.30     4.9       16          25   

    Type_Large  Type_Midsize  Type_Small  Type_Sporty  Type_Van  
6            0             1           0            0         0  
7            1             0           0            0         0  
8            1             0           0            0         0  
9            0             1           0            0         0  
10           1             0           0            0         0

In [42]:
from sklearn.linear_model import LinearRegression #線形回帰クラスのインポート
reg = LinearRegression() # 線形回帰クラスのインスタンス生成
reg.fit(X, y)           # 学習
yhat = reg.predict(X)   # 予測

In [43]:
print("y-切片= ",reg.intercept_)
print("係数 = ",reg.coef_)
print("決定変数= ",reg.score(X,y)) #決定係数の別計算

y-切片=  0.28729750255056885
係数 =  [ 0.58605175  0.41465523 -0.17139652  0.25820268 -0.04173108 -0.00421374
 -0.00147412 -0.0106137  -0.00927022 -0.00828079 -0.01827011 -0.00393354]
決定変数=  0.9999921253827208


### 問題

教科書79ページのtipsデータに対して線形回帰を用いてもらえるチップの額を予測せよ．

In [47]:
import seaborn as sns
tips = sns.load_dataset('tips')
tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [48]:
tips = pd.get_dummies(tips, drop_first=True) #ダミー変数の最初のものを除く
tips.head()

total_bill   tip  size  sex_Female  smoker_No  day_Fri  day_Sat  day_Sun  \
0       16.99  1.01     2           1          1        0        0        1   
1       10.34  1.66     3           0          1        0        0        1   
2       21.01  3.50     3           0          1        0        0        1   
3       23.68  3.31     2           0          1        0        0        1   
4       24.59  3.61     4           1          1        0        0        1   

   time_Dinner  
0            1  
1            1  
2            1  
3            1  
4            1

In [49]:
y = tips.tip                # 従属変数(price)の抽出
X = tips.drop('tip',axis=1) # 独立変数（特徴ベクトル）をpriceの列を除くことによって生成
X.head()

total_bill  size  sex_Female  smoker_No  day_Fri  day_Sat  day_Sun  \
0       16.99     2           1          1        0        0        1   
1       10.34     3           0          1        0        0        1   
2       21.01     3           0          1        0        0        1   
3       23.68     2           0          1        0        0        1   
4       24.59     4           1          1        0        0        1   

   time_Dinner  
0            1  
1            1  
2            1  
3            1  
4            1

In [50]:
from sklearn.linear_model import LinearRegression #線形回帰クラスのインポート
reg = LinearRegression() # 線形回帰クラスのインスタンス生成
reg.fit(X, y)           # 学習
yhat = reg.predict(X)   # 予測

In [51]:
print("y-切片= ",reg.intercept_)
print("係数 = ",reg.coef_)
print("決定変数= ",reg.score(X,y)) #決定係数の別計算

y-切片=  0.5908374259513769
係数 =  [ 0.09448701  0.175992    0.03244094  0.08640832  0.1622592   0.04080082
  0.13677854 -0.0681286 ]
決定変数=  0.47007812322060794


### B.3  ロジスティック回帰

iris（アヤメ）のデータを用いてロジスティック回帰（分類）を行う．


In [6]:
import pandas as pd

In [7]:
iris = pd.read_csv("http://logopt.com/data/iris.data",
                   names=['sepal length','sepal width','petal length','petal width', 'class'])

In [8]:
#  独立変数（特徴ベクトル） X 
X = iris[ ['sepal length','sepal width','petal length','petal width'] ]

In [9]:
#従属変数 y 
y = iris['class']
y.head() 

0    Iris-setosa
1    Iris-setosa
2    Iris-setosa
3    Iris-setosa
4    Iris-setosa
Name: class, dtype: object

In [10]:
#特徴ベクトルの抽出
X= iris[ ['sepal length','sepal width','petal length','petal width'] ]
X .head() 

sepal length  sepal width  petal length  petal width
0           5.1          3.5           1.4          0.2
1           4.9          3.0           1.4          0.2
2           4.7          3.2           1.3          0.2
3           4.6          3.1           1.5          0.2
4           5.0          3.6           1.4          0.2

## 手順１：分類するためのクラスをインポートして，インスタンスを生成する．

> ### 以下の例では，ロジスティック回帰を用いた分類を行う．

## 手順２：fitメソッドを用いて，データから学習する．

## 手順３：predictメソッドを用いて予測を行う．


In [11]:
from sklearn.linear_model import LogisticRegression #ロジスティック回帰クラスの読み込み

In [12]:
logreg =  LogisticRegression() # インスタンスの生成
logreg.fit(X, y) # 学習

/Users/okubo_yui/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/okubo_yui/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [14]:
logreg.predict([[3, 5, 4, 2]]) # 試しに予測

array(['Iris-virginica'], dtype=object)

### 予測と実際の誤差を検証（元データXを入れたときの予測y_predと本当の値yの比較）

### metricsにある正確度を計算する関数 accuracy_score を利用

In [15]:
y_pred = logreg.predict(X)

In [16]:
from sklearn import metrics
print(metrics.accuracy_score(y, y_pred))

0.96


### 問題

教科書81ページのtitanicデータに対してロジスティック回帰を行い，死亡確率の推定を行え．

ヒント：このデータは欠損値を含んでいる．pandasのところで学んだ欠損値処理を参照せよ．

また，このデータはターゲットとなる**survived**と同じ意味をもつ列を含んでいる．それを入れると正確度は当然100%になる．
その行を見つけて除くにはどうしたら良いか考えよ．

In [50]:
titanic = pd.read_csv('http://logopt.com/data/titanic.csv',index_col=0)
titanic1=titanic.dropna()
titanic1

Survived  Pclass  \
PassengerId                     
2                   1       1   
4                   1       1   
7                   0       1   
11                  1       3   
12                  1       1   
22                  1       2   
24                  1       1   
28                  0       1   
53                  1       1   
55                  0       1   
63                  0       1   
67                  1       2   
76                  0       3   
89                  1       1   
93                  0       1   
97                  0       1   
98                  1       1   
103                 0       1   
111                 0       1   
119                 0       1   
124                 1       2   
125                 0       1   
137                 1       1   
138                 0       1   
140                 0       1   
149                 0       2   
152                 1       1   
171                 0       1   
175                 0       1   
178                 0       1   
...               ...     ...   
738                 1       1   
742                 0       1   
743                 1       1   
746                 0       1   
749                 0       1   
752                 1       3   
760                 1       1   
764                 1       1   
766                 1       1   
773                 0       2   
780                 1       1   
782                 1       1   
783                 0       1   
790                 0       1   
797                 1       1   
803                 1       1   
807                 0       1   
810                 1       1   
821                 1       1   
824                 1       3   
836                 1       1   
854                 1       1   
858                 1       1   
863                 1       1   
868                 0       1   
872                 1       1   
873                 0       1   
880                 1       1   
888                 1       1   
890                 1       1   

                                                          Name     Sex   Age  \
PassengerId                                                                    
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
7                                      McCarthy, Mr. Timothy J    male  54.0   
11                             Sandstrom, Miss. Marguerite Rut  female   4.0   
12                                    Bonnell, Miss. Elizabeth  female  58.0   
22                                       Beesley, Mr. Lawrence    male  34.0   
24                                Sloper, Mr. William Thompson    male  28.0   
28                              Fortune, Mr. Charles Alexander    male  19.0   
53                    Harper, Mrs. Henry Sleeper (Myna Haxtun)  female  49.0   
55                              Ostby, Mr. Engelhart Cornelius    male  65.0   
63                                 Harris, Mr. Henry Birkhardt    male  45.0   
67                                Nye, Mrs. (Elizabeth Ramell)  female  29.0   
76                                     Moen, Mr. Sigurd Hansen    male  25.0   
89                                  Fortune, Miss. Mabel Helen  female  23.0   
93                                 Chaffee, Mr. Herbert Fuller    male  46.0   
97                                   Goldschmidt, Mr. George B    male  71.0   
98                             Greenfield, Mr. William Bertram    male  23.0   
103                                  White, Mr. Richard Frasar    male  21.0   
111                             Porter, Mr. Walter Chamberlain    male  47.0   
119                                   Baxter, Mr. Quigg Edmond    male  24.0   
124                                        Webber, Miss. Susan  female  32.5   
125                                White, Mr. Percival Wayland    male  54.0   
137          

In [51]:
#  独立変数（特徴ベクトル） X 
X = titanic1[ ['Age','Fare' ]]
X.head()

Age     Fare
PassengerId               
2            38.0  71.2833
4            35.0  53.1000
7            54.0  51.8625
11            4.0  16.7000
12           58.0  26.5500

In [52]:
#従属変数 y 
y = titanic1['Survived']
y.head() 

PassengerId
2     1
4     1
7     0
11    1
12    1
Name: Survived, dtype: int64

In [53]:
from sklearn.linear_model import LogisticRegression #ロジスティック回帰クラスの読み込み
logreg =  LogisticRegression() # インスタンスの生成
logreg.fit(X, y) # 学習

/Users/okubo_yui/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [54]:
logreg.predict([[31,52]]) # 試しに予測

array([1])

In [55]:
y_pred = logreg.predict(X)
from sklearn import metrics
print(metrics.accuracy_score(y, y_pred))

0.7103825136612022


### 問題
 "http://logopt.com/data/cancer.csv" にある胸部癌か否かを判定するデータセットを用いて分類を行え．

最初の列**diagnosis**が癌か否かを表すものであり，'M'が悪性（malignant），'B'が良性（benign）

In [56]:
cancer = pd.read_csv("http://logopt.com/data/cancer.csv", index_col=0)
#従属変数 y 
y = cancer['diagnosis']
#特徴ベクトルの抽出
X= cancer.iloc[ : , 1: ]
X .head() 

radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                               
842302          17.99         10.38          122.80     1001.0   
842517          20.57         17.77          132.90     1326.0   
84300903        19.69         21.25          130.00     1203.0   
84348301        11.42         20.38           77.58      386.1   
84358402        20.29         14.34          135.10     1297.0   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
842302            0.11840           0.27760          0.3001   
842517            0.08474           0.07864          0.0869   
84300903          0.10960           0.15990          0.1974   
84348301          0.14250           0.28390          0.2414   
84358402          0.10030           0.13280          0.1980   

          concave points_mean  symmetry_mean  fractal_dimension_mean  \
id                                                                     
842302                0.14710         0.2419                 0.07871   
842517                0.07017         0.1812                 0.05667   
84300903              0.12790         0.2069                 0.05999   
84348301              0.10520         0.2597                 0.09744   
84358402              0.10430         0.1809                 0.05883   

                   ...             radius_worst  texture_worst  \
id                 ...                                           
842302             ...                    25.38          17.33   
842517             ...                    24.99          23.41   
84300903           ...                    23.57          25.53   
84348301           ...                    14.91          26.50   
84358402           ...                    22.54          16.67   

          perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
id                                                                           
842302             184.60      2019.0            0.1622             0.6656   
842517             158.80      1956.0            0.1238             0.1866   
84300903           152.50      1709.0            0.1444             0.4245   
84348301            98.87       567.7            0.2098             0.8663   
84358402           152.20      1575.0            0.1374             0.2050   

          concavity_worst  concave points_worst  symmetry_worst  \
id                                                                
842302             0.7119                0.2654          0.4601   
842517             0.2416                0.1860          0.2750   
84300903           0.4504                0.2430          0.3613   
84348301           0.6869                0.2575          0.6638   
84358402           0.4000                0.1625          0.2364   

          fractal_dimension_worst  
id                                 
842302                    0.11890  
842517                    0.08902  
84300903                  0.08758  
84348301                  0.17300  
84358402                  0.07678  

[5 rows x 30 columns]

In [57]:
from sklearn.linear_model import LogisticRegression #ロジスティック回帰クラスの読み込み
logreg =  LogisticRegression() # インスタンスの生成
logreg.fit(X, y) # 学習

/Users/okubo_yui/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [64]:
logreg.predict([[19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,23.57,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,0.2,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0,4]]) # 試しに予測

array(['M'], dtype=object)

In [65]:
y_pred = logreg.predict(X)
from sklearn import metrics
print(metrics.accuracy_score(y, y_pred))

0.9595782073813708


## 問題

$K$ 近傍法とは，(Xの空間で）データに近い $K$ 個のデータの値(y)の多数決で分類を行う最も簡単な分類手法である．

以下のようにしてクラスをインポートできる．
```python
from sklearn.neighbors import KNeighborsClassifier
```

クラス **KNeighborsClassifier** は引数 **n_neighbors** でパラメータ $K$ を設定できる．

irisデータに対して，$K$ 近傍法で元データを予測したしたときの正確度を計算せよ．パラメータ $K$（近傍の数）が5のときはどうか？また，パラメータ $K$ が 1のときはどうか？

In [47]:
from sklearn.neighbors import KNeighborsClassifier

### B.4  正規化とB.7 仮説の評価

### 多項式回帰

宣伝の効果のデータを用いて多項式回帰を行う．

pandasで2次の項（たとえばテレビとラジオの相乗効果の列 TV*Radio）を生成してから，線形回帰を行う．

In [38]:
data = pd.read_csv('http://logopt.com/data/Advertising.csv', index_col =0)
data['TV*Radio'] = data.TV * data.Radio
data['TV*Newspaper'] = data.TV * data.Newspaper
data['Radio*Newspaper'] = data.Radio * data.Newspaper
data.head()

TV  Radio  Newspaper  Sales  TV*Radio  TV*Newspaper  Radio*Newspaper
1  230.1   37.8       69.2   22.1   8697.78      15922.92          2615.76
2   44.5   39.3       45.1   10.4   1748.85       2006.95          1772.43
3   17.2   45.9       69.3    9.3    789.48       1191.96          3180.87
4  151.5   41.3       58.5   18.5   6256.95       8862.75          2416.05
5  180.8   10.8       58.4   12.9   1952.64      10558.72           630.72

In [39]:
X = data[['TV', 'Radio', 'Newspaper','TV*Radio','TV*Newspaper','Radio*Newspaper']]
y = data['Sales']

In [40]:
from sklearn.linear_model import LinearRegression #クラスのインポート
lin_reg = LinearRegression()     # 線形回帰クラスのインスタンス生成
lin_reg.fit(X, y)    # 学習
yhat = lin_reg.predict(X)   # 予測
print(lin_reg.score(X,y)) #決定係数　（線形回帰だと R^2は 0.897210638179

0.9686311066255571


In [41]:
print("y-切片= ",lin_reg.intercept_)
print("係数 = ",lin_reg.coef_)

y-切片=  6.460158489941863
係数 =  [ 2.03270962e-02  2.29291894e-02  1.70339370e-02  1.13928001e-03
 -7.97143541e-05 -1.09597607e-04]


### 過剰適合（過学習）

元データを入れてテストをすることは良くない．

### irisデータに対してK(=1)近傍法を適用したときのように，過剰適合をしてしまう！

### 学習データをテストデータを分ける！ train_test_split関数を使うと簡単．

**（最近のバージョンではmodel_selection内にあるが，古いバージョンの場合には,cross_validationの中にあるので注意）**

例としてBostonの住宅データの予測を線形回帰とリッジ回帰（正規化を入れた回帰）行う．

medvが住宅の価格で，他のデータ（犯罪率や人口など）から予測する．

In [68]:
boston = pd.read_csv('http://logopt.com/data/Boston.csv',index_col=0)
boston.head()

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
1  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
2  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
3  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
4  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
5  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

    black  lstat  medv  
1  396.90   4.98  24.0  
2  396.90   9.14  21.6  
3  392.83   4.03  34.7  
4  394.63   2.94  33.4  
5  396.90   5.33  36.2

In [43]:
X = boston.iloc[:,range(13)] # 最後の列以外のデータを独立変数（特徴ベクトル）として抽出
y = boston.medv #最後の列(medv)を従属変数として抽出

In [44]:
from sklearn.model_selection import train_test_split #古いバージョンの場合には,cross_validationの中にある．
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)#30%のデータをテスト用，それ以外を学習用に分離

In [45]:
from sklearn.linear_model import LinearRegression #クラスのインポート
lin_reg = LinearRegression()     # 線形回帰クラスのインスタンス生成
lin_reg.fit(X_train, y_train)    # 学習
yhat = lin_reg.predict(X_test)   # 予測
print(lin_reg.score(X_test,y_test)) #決定係数

0.7435699440123182


### リッジ回帰

教科書460ページになある正規化を追加した線形回帰は，**リッジ回帰**とよばれる．

パラメータ $\lambda$ はリッジ回帰クラス Ridge では引数 alpha で与える．


In [46]:
from sklearn.linear_model import Ridge #クラスのインポート
ridge_reg = Ridge(alpha=100.0)     # リッジ回帰クラスのインスタンス生成
ridge_reg.fit(X_train, y_train)    # 学習
yhat = ridge_reg.predict(X_test)   # 予測
print(ridge_reg.score(X_test,y_test)) #決定係数

0.7230626569260372


### 問題

1. irisデータを学習用とテスト用に分けてから，ロジスティック回帰による分類を行い，正確度を計算せよ．
2. titanicデータ学習用とテスト用に分けてから，ロジスティック回帰による分類を行い，正確度を計算せよ．
3. 例題2で用いた ダイアモンドの価格データhttp://logopt.com/data/Diamond.csv に対して線形回帰とリッジ回帰による予測を行え．また，学習用とテスト用に分けて決定変数を計算し，評価せよ．


In [83]:
iris = pd.read_csv("http://logopt.com/data/iris.data",
                   names=['sepal_length','sepal_width','petal_length','petal_width', 'class'],index_col=0)
iris.head()

sepal_width  petal_length  petal_width        class
sepal_length                                                     
5.1                   3.5           1.4          0.2  Iris-setosa
4.9                   3.0           1.4          0.2  Iris-setosa
4.7                   3.2           1.3          0.2  Iris-setosa
4.6                   3.1           1.5          0.2  Iris-setosa
5.0                   3.6           1.4          0.2  Iris-setosa

In [92]:
X = iris.iloc[:,range(2)] # 最後の列以外のデータを独立変数（特徴ベクトル）として抽出
y = iris.petal_width 

In [90]:
from sklearn.model_selection import train_test_split #古いバージョンの場合には,cross_validationの中にある．
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.6)#30%のデータをテスト用，それ以外を学習用に分離

In [91]:
from sklearn.linear_model import LinearRegression #クラスのインポート
lin_reg = LinearRegression()     # 線形回帰クラスのインスタンス生成
lin_reg.fit(X_train, y_train)    # 学習
yhat = lin_reg.predict(X_test)   # 予測
print(lin_reg.score(X_test,y_test)) #決定係数

0.9096055479553098


### B.5 カーネルとSVM (Support Vector Machine) 

カーネルはクラス **SVC** (Support Vector Classifier) の引数 **kernel**で与える．

Gaussカーネル（引数は'rbf'）を用いると非線形な境界をもつ問題に対しても，精度良く分類が可能になる．

irisデータで試してみよう．

In [93]:
import pandas as pd
iris = pd.read_csv("http://logopt.com/data/iris.data",
                   names=['sepal length','sepal width','petal length','petal width', 'class'])
X = iris[ ['sepal length','sepal width','petal length','petal width'] ]
y = iris['class']
from sklearn.model_selection import train_test_split  #古いバージョンの場合には,cross_validationの中にある．
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

In [94]:
from sklearn.svm import SVC      #クラスのインポート
svc = SVC(kernel='rbf', gamma="scale")        # インスタンス生成
#svc = SVC(kernel='linear') 
svc.fit(X_train, y_train)    # 学習
yhat = svc.predict(X_test)   # 予測
from sklearn import metrics
print(metrics.accuracy_score(y_test, yhat)) # 精度

0.9777777777777777


### B.7 仮説の評価

###  １つのテストデータだけではばらつきが大きすぎて信用できない！

### 交差検証を行えば良い！cross_val_scoreを用いる．

上で用いたサポートベクトル分類 svcのインスタンスで試してみる．

引数の*cv*は交差検証(cross validiation)の分割数．

引数の**scoring='accuracy'**は分類の正確度を表す指標を指定している．

（回帰分析のときには，引数を決定変数'r2'などに設定する．）


In [95]:
from sklearn.model_selection import cross_val_score #古いバージョンの場合には,cross_validationの中にある．
scores = cross_val_score(svc, X, y, cv=10, scoring='accuracy')
print(scores)
print(scores.mean())

[1.         0.93333333 1.         1.         1.         0.93333333
 0.93333333 1.         1.         1.        ]
0.9800000000000001


### B.8 ニューラルネット，B.9 単純Bayes，B.10 決定木，B.11 アンサンブル法（ランダム森）

他の手法も，scikit-learnでは同じ手順で分析を行うことができる．

上の続きとしてirisデータで色々試してみよう．

ニューラルネットは旧バージョンでは動かないが，
http://playground.tensorflow.org/
で深層学習（多階層のニューラルネット）を体験することができる．


In [96]:
from sklearn.neural_network import MLPClassifier #ニューラルネット（旧バージョンでは実装されていないので，跳ばして良い．）
neural = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(10, 3), random_state=1)
neural.fit(X_train, y_train)    # 学習
yhat = neural.predict(X_test)   # 予測
from sklearn import metrics
print(metrics.accuracy_score(y_test, yhat)) # 精度　
scores = cross_val_score(neural, X, y, cv=10, scoring='accuracy')
print(scores)
print(scores.mean())

0.9777777777777777
[1.         0.93333333 1.         1.         0.93333333 0.93333333
 0.93333333 0.93333333 1.         1.        ]
0.9666666666666668


In [97]:
from sklearn.naive_bayes import GaussianNB #単純Bayes
bayes = GaussianNB()
bayes.fit(X_train, y_train)    # 学習
yhat = bayes.predict(X_test)   # 予測
from sklearn import metrics
print(metrics.accuracy_score(y_test, yhat)) # 精度　
scores = cross_val_score(bayes, X, y, cv=10, scoring='accuracy')
print(scores)
print(scores.mean())

0.9555555555555556
[0.93333333 0.93333333 1.         0.93333333 0.93333333 0.93333333
 0.86666667 1.         1.         1.        ]
0.9533333333333334


In [98]:
from sklearn import tree
tree_class = tree.DecisionTreeClassifier() #決定木
tree_class.fit(X_train, y_train)    # 学習
yhat = tree_class.predict(X_test)   # 予測
from sklearn import metrics
print(metrics.accuracy_score(y_test, yhat)) # 精度　
scores = cross_val_score(tree_class, X, y, cv=10, scoring='accuracy')
print(scores)
print(scores.mean())

0.9333333333333333
[1.         0.93333333 1.         0.93333333 0.93333333 0.86666667
 0.93333333 0.93333333 1.         1.        ]
0.9533333333333334


In [99]:
from sklearn.ensemble import RandomForestClassifier #ランダム森
forest = RandomForestClassifier()
forest.fit(X_train, y_train)    # 学習
yhat = forest.predict(X_test)   # 予測
from sklearn import metrics
print(metrics.accuracy_score(y_test, yhat)) # 精度　
scores = cross_val_score(forest , X, y, cv=10, scoring='accuracy')
print(scores)
print(scores.mean())


0.9555555555555556
[1.         0.93333333 1.         0.93333333 0.86666667 0.93333333
 0.93333333 1.         1.         1.        ]
0.96


/Users/okubo_yui/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


### 問題

titanicデータ学習用とテスト用に分けてから，ニューラルネット（旧バージョンでは実装されていないので，跳ばして良い．），単純Bayes，決定木，ランダム森による分類を行い，交差検証を行え．

### B.12 クラスタリング

UCI機械学習レポジトリのワインに関するデータセットを用いてクラスタリングを解説する．

使用するのは**KMeans**クラスで実装されている $k$-平均法である．

元データはこちらに格納されている．

http://logopt.com/data/wine.data

列名は https://archive.ics.uci.edu/ml/datasets/Wine で解説されているが，必要ならば以下のリストを用いて，列名を設定して読み込め．

L = [ 'Alcohol', 'Malic','Ash', 'Alcalinity', 'Magnesium', 'Phenols', 'Flavanoids', 'Nonflavanoid', 'Proanthocyanins', 'Color', 'Hue', 'OD280', 'OD315', 'Proline']

In [100]:
import pandas as pd
L = [ 'Alcohol', 'Malic','Ash', 'Alcalinity', 'Magnesium', 'Phenols', 'Flavanoids', 'Nonflavanoid',
     'Proanthocyanins', 'Color', 'Hue', 'OD280', 'OD315', 'Proline']
wine = pd.read_csv('http://logopt.com/data/wine.data',names=L)
wine.head()

Alcohol  Malic   Ash  Alcalinity  Magnesium  Phenols  Flavanoids  \
0        1  14.23  1.71        2.43       15.6      127        2.80   
1        1  13.20  1.78        2.14       11.2      100        2.65   
2        1  13.16  2.36        2.67       18.6      101        2.80   
3        1  14.37  1.95        2.50       16.8      113        3.85   
4        1  13.24  2.59        2.87       21.0      118        2.80   

   Nonflavanoid  Proanthocyanins  Color   Hue  OD280  OD315  Proline  
0          3.06             0.28   2.29  5.64   1.04   3.92     1065  
1          2.76             0.26   1.28  4.38   1.05   3.40     1050  
2          3.24             0.30   2.81  5.68   1.03   3.17     1185  
3          3.49             0.24   2.18  7.80   0.86   3.45     1480  
4          2.69             0.39   1.82  4.32   1.04   2.93      735

In [101]:
from sklearn.cluster import KMeans  # クラスをインポート
kmeans = KMeans(n_clusters=8)       # インスタンス生成
kmeans.fit(wine)                    # 学習

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
    n_clusters=8, n_init=10, n_jobs=None, precompute_distances='auto',
    random_state=None, tol=0.0001, verbose=0)

In [102]:
wine['label'] = kmeans.labels_
wine.head()

Alcohol  Malic   Ash  Alcalinity  Magnesium  Phenols  Flavanoids  \
0        1  14.23  1.71        2.43       15.6      127        2.80   
1        1  13.20  1.78        2.14       11.2      100        2.65   
2        1  13.16  2.36        2.67       18.6      101        2.80   
3        1  14.37  1.95        2.50       16.8      113        3.85   
4        1  13.24  2.59        2.87       21.0      118        2.80   

   Nonflavanoid  Proanthocyanins  Color   Hue  OD280  OD315  Proline  label  
0          3.06             0.28   2.29  5.64   1.04   3.92     1065      4  
1          2.76             0.26   1.28  4.38   1.05   3.40     1050      4  
2          3.24             0.30   2.81  5.68   1.03   3.17     1185      1  
3          3.49             0.24   2.18  7.80   0.86   3.45     1480      5  
4          2.69             0.39   1.82  4.32   1.04   2.93      735      2

### 問題

irisのデータセットの各データを $k$-平均法を用いて3つのクラスターに分けよ．

### B.13 主成分分析

irisデータセットを用いて主成分分析と可視化の方法を説明する．


In [103]:
import pandas as pd
iris = pd.read_csv("http://logopt.com/data/iris.data",
                   names=['sepal length','sepal width','petal length','petal width', 'class'])
X = iris[ ['sepal length','sepal width','petal length','petal width'] ]

In [104]:
from sklearn.decomposition import PCA
pca = PCA(n_components=2) # 2次元に射影
pca.fit(X)
pca.components_ #射影行列

array([[ 0.36158968, -0.08226889,  0.85657211,  0.35884393],
       [ 0.65653988,  0.72971237, -0.1757674 , -0.07470647]])

In [105]:
Z = pca.transform(X) # 射影したデータ
# 射影した2次元データを元のデータフレームに追加
iris['X']=Z[:,0]
iris['Y']=Z[:,1]

In [ ]:
#seabornの lmplot で描画（hueは色調指定，fit_reg=Falseは線を描画しない）
import seaborn as sns
%matplotlib inline
sns.lmplot(x='X',y='Y',hue='class',fit_reg=False,data=iris);

### 問題

http://logopt.com/data/drinks.csv にある国別のアルコール摂取量データを用いて主成分分析を行え．

4次元の数値データを2次元に射影し，2次元座標で表示せよ．その際，色調としては，大陸(continent)列を用いよ．

### 問題

クラスタリングの例で用いたワインのデータを2次元に射影して，クラスタリングされたラベルを色調として描画せよ．